### DISCUSSION TITLE
Saleor SDK and Auth0

### DISCUSSION THREAD

**Discussion by berthj193** [saleor/saleor#7829] [Q&A]
What's the best way to make use of `@saleor/sdk` and `Auth0` in combination without (or with minimal) changes upon Saleor updates?

Our current platform consists of:
- Backend:
  - Saleor core API
  - A few more standalone PHP driven APIs
- Frontend:
  - `saleor-storefront@3.0.0-a.5`
  - `@saleor/sdk@0.3.0-10`
  - `@auth0/auth0-react` (interim)

I have been trying to integrate Auth0 as an authentication system and I faced some blockers.
- No issues found creating an OIDC plugin in Saleor core.
- `@saleor/sdk` doesn't provide external OIDC login until version `0.4.0-2`
- `saleor-dashboard` supports OIDC but the logic is highly coupled into dashboard code itself.
- removed @saleor/sdk from npm dependency
- patched @saleor/sdk code into storefront project folder and made changes to Auth API to support external login.
- used `@auth0/auth0-react` as a temporary solution to obtain & manage ID / access / refresh tokens. But it was difficult to relay authentication status to saleor SDK.

Leaning towards to:
-  Use Saleor as a resource server which only validates tokens
   - https://docs.saleor.io/docs/3.0/developer/available-plugins/openid-connect/#saleor-as-a-client-for-authorization-server
   - https://docs.saleor.io/docs/3.0/developer/available-plugins/openid-connect/#saleor-as-a-resource-server
- Start a new next.js project which is similar to `saleor-dashboard`
- Create own graphql layer

Question:
- How stable is API of `@saleor/sdk@0.4.0-2`?
- What potential issues can I encounter if I integrate an well-maintained official Auth0 SDK like `@auth0/auth0-react?
- How to synchronize authentication status between Saleor and Auht0 SDK?

**Comment by patrys** [comment]
Unless you have a good reason not to do that, I suggest that you use the Auth0 JS SDK and use the resulting access token to sign requests going to Saleor. In that case, you'll need to implement `authenticate_user` in your Saleor plugin to verify the token against the JWKS provided by the Auth0 server and use the email address to select the correct account.

PS: Our Cloud service offers an OpenID Connect Saleor plugin that works with Auth0.

**Reply by berthj193** [reply]
Thanks for your quick answer!

Yea, we have already implemented OpenID Connect plugin on Saleor Core partially.

Just wondering what's the path for frontend? We still need to use Saleor features like Checkout along with Auth0 authentication.

Assuming Saleor JS SDK doesn't provide OIDC yet, do you suggest
- fork/patch Saleor SDK and make changes accordingly? Modify [something like this](https://github.com/mirumee/saleor-sdk/blob/f5a1fa8d1d7498100df72607b519114d37304be6/src/core/auth.ts#L143) into external token verify?
- or follow [Saleor Dashboard](https://github.com/mirumee/saleor-dashboard/blob/master/src/auth/hooks/useExternalAuthProvider.ts)? (which doesn't rely on Saleor SDK, but has its own external authentication layer which talks with Saleor Core plugin)

**Reply by patrys** [reply]
Paging @jwm0 📟

**Reply by marianmazarovici** [reply]
> PS: Our Cloud service offers an OpenID Connect Saleor plugin that works with Auth0.

Hi @patrys 
VP of eng for vonq.com here.
In reply to your remark, we'd love to use Saleor Cloud but that's not possible, for several reasons.
At vonq.com, we are building a B2B enterprise-focused E-commerce solution to replace our legacy stack. 
Although Saleor's Apps and Plugins architecture is a brilliant way of extending Saleor, some of our B2B acceptance criteria require some subtle, but fundamental customisations to the User, Product and Order models. Using apps would require heavy custom development and extra architectural complexity which defeats the whole purpose of leveraging Saleor. 

Happy to discuss if you're up for it. If it turns out there's something beneficial for Saleor in what we're building internally, we are delighted to contribute back to the open-source community.

**Reply by jwm0** [reply]
@mingthegit we're working on adding the OIDC method to SDK v0.4 very soon, so you might not need to fork it unless you need it now. After we release first stable revamped SDK, we'll be integrating it into Dashboard.

**Reply by berthj193** [reply]
@jwm0 That's a great news!

Just curious if it will be possible to have [@auth0/auth0-react](https://github.com/auth0/auth0-react) plugin as an authorization center.
- `auth0-react` plugin obtain tokens from Auth0 directly and those tokens are injected into Saleor SDK's auth client upon successful login. Following API calls for Saleor Core will use this token. (I saw `setAuthToken` no longer exists)
- it can clear out authentication state of Saleor SDK upon `auth0-react`'s logout action.

As I posted in the original question, I have checked how [Saleor Dashboard](https://github.com/mirumee/saleor-dashboard/blob/master/src/auth/hooks/useExternalAuthProvider.ts) implemented OIDC without Saleor SDK.
I'm fine with that approach too, assuming that's the one you're adddin to SDK v0.4. I can pull token from Saleor SDK and use that for other API calls.

But practically for our B2B platform where we have multiple external protected API services (along with Saleor Core),
- I'm leaning towards to
https://docs.saleor.io/docs/3.0/developer/available-plugins/openid-connect/#saleor-as-a-resource-server
- instead of 
https://docs.saleor.io/docs/3.0/developer/available-plugins/openid-connect/#saleor-as-a-client-for-authorization-server

**Reply by patrys** [reply]
> Although Saleor's Apps and Plugins architecture is a brilliant way of extending Saleor, some of our B2B acceptance criteria require some subtle, but fundamental customisations to the User, Product and Order models. Using apps would require heavy custom development and extra architectural complexity which defeats the whole purpose of leveraging Saleor.

@marianmazarovici I'd love to hear how we can make this solvable using metadata/private metadata/attributes.

PS: We're planning to extend attribute support to orders and users.

**Reply by berthj193** [reply]
@jwm0 Hi Jakub, do you have a brief roadmap / timeline you can share? I see you are making good progress with new versions (`0.4.0-3` latest), just wondering when you plan to release OIDC authentication.

**Reply by haneef-photon** [reply]
@mingthegit were you able to get things working? I took a look at the latest SDK, from what I was able to understand is, for the OIDC resource server flow, these are the steps that are required: 
1. At the client (storefront), need to use the Auth, `externalVerify()` method to pass the pluginID, token to the server.
2. Create a new plugin and implement the `user_authenticate()` method which accepts the token, parse and verify the token and return the `User` object associated with the token.

Hopefully this is all that is required from implementation part for the resource server flow. Feel free to correct me if I missed something and did not understand the requirements for this implementation correctly.

**Reply by toastedtoast** [reply]
@haneef-photon 
Hi, have you been able to use auth0 with a self hosted saleor core?
Actually I need some guidance how to approach the plugin implementation best. Can anyone share good resources?

Just need to implement the authenticate_user flow.

**Reply by haneef-photon** [reply]
@toastedtoast I could not make any headway, from what I understood we need to create our own plugin but I do not have enough python skills to work on it.

**Comment by patrys** [comment]
We're open-sourcing the OIDC plugin here: #9406

**Reply by marianmazarovici** [reply]
Thank you so much! Just in time
